`WCOrderShipmentProviderModel` to `Room`

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Closes: [AINFRA-555](https://linear.app/a8c/issue/AINFRA-555/migrate-wcordershipmentprovidermodel)

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCOrderShipmentProviderModel` to `Room`.

Additionally, it addressed a bug of not updating shipment providers after initial fetch. This was needed to correctly test the migration. Commit with fix: d0b41c3

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

You can reproduce the issue by following all test steps on `trunk`. In point 7. of the `Updating carriers` section you won't see added providers.

### Testing information

#### Basic fetch
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
1. Prepare a cup of coffee or tea as it'll have multiple steps.
2. On your test site install `Shipment Tracking`. Make sure this is plugin **authored by WooCommerce**. You can get the plugin by logging in to `https://woocommerce.com/` using your `@automattic.com` mail address and going to `Downloads` section. Enable the plugin.
3. Make sure your test store has a physical product.
4. In the mobile app, create a new order: include a physical product there. Tap on "Collect payment" on the same screen and select "Cash"
5. Mark order as completed
6. Open the order
7. In **CUSTOMER** section, you should see **ADD TRACKING** button. Click on it.
8. Tap `Carrier`, and see that you can see the list of carriers, including `Custom Carrier` option (which will be visible before the rest is fetched, I think it's ok)

#### Updating carriers

Then, starting from the last step of above test case:
1. Go to your test website
2. Go to `Tools` > `Theme File Editor` 
3. On the right sidebar tap on `functions.php`
4. Paste the snippet below: it'll add new providers
5. Tap on `Update file`
6. Go to the mobile app, perform all steps needed to select providers
7. Assert you can see "France | French New Mail" and "United Kingdom | UPS" providers
8. You can also revert the change made to `funcations.php`, and open the mobile app again to see if the above got removed.

<details><summary>Snippet to add new providers </summary>
<p>

from docs:

```php
/**
 * wc_shipment_tracking_add_custom_provider
 *
 * Adds custom provider to shipment tracking or add existing providers to another country.
 * 
 * Change the country name, the provider name, and the URL (it must include the %1$s)
 * Add one provider per line
*/

add_filter( 'wc_shipment_tracking_get_providers' , 'wc_shipment_tracking_add_custom_provider' );

function wc_shipment_tracking_add_custom_provider( $providers ) {
	
	$providers['France']['French New Mail'] = 'https://example.com?id=%1$s'; // add a custom provider
	$providers['United Kingdom']['UPS'] = 'https://www.ups.com/track?loc=en_GB&tracknum=%1$s&requester=WT'; // add UPS to the United Kingdom
	return $providers;
}
```

</p>
</details> 


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="1471" alt="Screenshot 2025-06-21 at 14 33 49" src="https://github.com/user-attachments/assets/4423b2bf-5a26-428b-9ff3-0c85aa2c570a" />


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Migrate `WCOrderShipmentProviderModel` from WellSql to Room. Provide migrations.

wzieba committed:
Adjust unit tests after WCOrderShipmentProviderModel migration

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Merge branch 'trunk' into migrate_shipment_providers

wzieba committed:
Provide database migration

wzieba committed:
Fix: always fetch shipment providers, even if cached

Ensure shipment providers are re-fetched from the backend, even when cached locally.

Previously, updates to shipment providers made via the web weren't reflected in the app unless it was reinstalled or its data was cleared. This change addresses that issue by always triggering a fetch, ensuring data consistency.

Additionally, this refactors the fetching logic toward a more CQRS-style approach:
- The `fetch` operation is now treated as a command, returning only a result.
- Actual data is exposed as a flow via Room (query), demonstrating the benefits of the ongoing Room migration.

wzieba committed:
Provide `OrderShipmentProvidersDao#replaceAll` instead of granular methods

The consumer of `OrderShipmentProvidersDao` only needs to replace data. This commits provides a transaction method for this purpose, encapsulating the logic.

+ removes unused methods

wzieba committed:
Minor code readability improvements

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/tracking/AddOrderTrackingProviderListViewModel.kt:
```diff
@@ -41,47 +45,61 @@ class AddOrderTrackingProviderListViewModel @Inject constructor(
     )
     private var trackingProviderListViewState by trackingProviderListViewStateData
 
-    private var providersList = emptyList<OrderShipmentProvider>()
-
```
Hotspot: removing state duplication (now this VM relies only on its `State` class)

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/tracking/OrderShipmentProvidersRepository.kt:
```diff
@@ -13,38 +16,47 @@ class OrderShipmentProvidersRepository @Inject constructor(
-    @Suppress("ReturnCount")
-    suspend fun fetchOrderShipmentProviders(orderId: Long): List<OrderShipmentProvider>? {
-        // Check db first
-        val providersInDb = getShipmentProvidersFromDB()
-        if (providersInDb.isNotEmpty()) {
-            return providersInDb
```
Hotspot: the culprit of the bug. If any providers were in DB, this method never requested providers for an update.

ParaskP7 replied:
Praise (❤️): Hmmm, this was indeed very interesting, I guess whoever wrote that code wasn't expecting this list or providers to ever change... 🤷

Great fix on all aspects @wzieba , from exposing a flow, to converting the list of `OrderShipmentProvider` into a result, which is much easier to reason about, and in general, the whole refactor + fix you did here. 🥇

wzieba committed:
Remove `WCOrderShipmentTrackingModel` from WellSql

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCOrderShipmentProviderModel.kt:
```diff
@@ -1,21 +1,15 @@
+
+@Entity(
+    tableName = "OrderShipmentProviderEntity",
+    primaryKeys = ["localSiteId", "carrierName", "country"]
```
Primay key doesn't matter that much, since we remove all data and add new one on update. Nevertheless, I think this set of fields is correct to uniquely describe a provider.

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt:
```diff
@@ -680,7 +676,7 @@ class WCOrderStore @Inject constructor(
         updateOrderStatusAndPaymentDetails(orderId, site, newStatus, null, null)
 
-    suspend fun updateOrderStatusAndPaymentDetails(
+    fun updateOrderStatusAndPaymentDetails(
```
>  it performs asynchronous database operations

No, it doesn't. It synchronously runs a coroutine. Making this method `suspend` is incorrect.

ParaskP7 commented:
👋 @wzieba !

> 2. On your test site install Shipment Tracking.

I got my cup of coffee and I am now stuck at this step, I don't know how to download `Shipment Tracking` for free, can you help? 🙏

PS: I tried adding a coupon for me so that I could apply that during the checkout process, but no luck, how did you do it? 🤔

wzieba commented:
@ParaskP7 👋 , could you please clarify on which step from

> You can get the plugin by logging in to https://woocommerce.com/ using your @automattic.com mail address and going to Downloads section. Enable the plugin.

you stuck?

ParaskP7 commented:
👋 @wzieba !

![image](https://github.com/user-attachments/assets/30499dc7-e73b-473d-9d9d-1f241fe2517d)

Does that help?

ParaskP7 commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt:
```diff
@@ -2198,6 +2198,10 @@ open class WellSqlConfig : DefaultWellConfig {
                 }
+
+                218 -> migrateAddOn(ADDON_WOOCOMMERCE, version) {
+                    db.execSQL("DROP TABLE IF EXISTS WCOrderShipmentTrackingModel")
```
Warning (⚠️): I still haven't done a full review @wzieba , but I noticed this here and wanted to flag it so that we get it out of our way asap: I think you meant to drop `WCOrderShipmentProviderModel` instead, right? 😊

wzieba replied:
Right, thanks! I'm surprised the app didn't crash - I guess it would when navigating to anything related to the actual tracking. Fixed in d1823d96aca087604a804c8c021a10246a3eecdb

ParaskP7 replied:
Thanks for the quick fix! 🚀 

> I guess it would when navigating to anything related to the actual tracking.

True! 😄

wzieba commented:
<img width="570" alt="Screenshot 2025-06-23 at 19 37 19" src="https://github.com/user-attachments/assets/354cd859-49ea-4192-a118-d13c64605a48" />

You have to navigate to "Downloads" section

wzieba committed:
Fix removing incorrect table from WellSql

ParaskP7 commented:
> You have to navigate to "Downloads" section

Yes, and from there I search `Shipment Tracking`.

wzieba commented:
<img width="1419" alt="Screenshot 2025-06-23 at 19 40 38" src="https://github.com/user-attachments/assets/e87e0dba-dd07-4988-8a25-2abaf8237e97" />

I think you might've missed this 🙂

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCOrderShipmentProviderModel.kt:
```diff
@@ -1,21 +1,15 @@
+    primaryKeys = ["localSiteId", "carrierName", "country"]
+)
+data class WCOrderShipmentProviderModel(
+    val localSiteId: LocalId,
```
Question (❓): Why did you chose to remove the default value here (`... = LocalId(0)`). I am asking that because for such entities we chose to keep that default value, both for local and remote ids, plus (I guess) for all the data class fields, just me wonder here what changed... 🤔 😊

wzieba replied:
I realized at some point that we shouldn't really provide default values for mandatory fields: one cannot save `WCOrderShipmentProviderModel` without a valid local site id, so having a default value is rather an issue (someone might by mistake forget to provide this and end up with assigning provider to incorrect site).

ParaskP7 replied:
Hmmm, this totally makes sense @wzieba, maybe we should create a follow-up PR at some point removing all those default values, see where that takes us, maybe we'll discovered a bug or two... 🤞

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/OrderShipmentProvidersDao.kt:
```diff
@@ -0,0 +1,39 @@
+    abstract fun observeOrderShipmentProviders(siteId: LocalId): Flow<List<WCOrderShipmentProviderModel>>
+
+    @Transaction
+    open suspend fun replaceAll(siteId: LocalId, providers: List<WCOrderShipmentProviderModel>) {
```
Praise (❤️): Nicely done, love it, much better and a great usage of `@Transaction`! 🥇

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected, an awesome job you did here! 🥇 x ❤️ ^ 🚀 

> Additionally, it addressed a bug of not updating shipment providers after initial fetch. This was needed to correctly test the migration. Commit with fix: https://github.com/woocommerce/woocommerce-android/commit/d0b41c3be6dda2f52f40b9adecc6363f4d273e86

Praise (❤️): Great job well done, thanks! 🚀

wzieba commented:
Thanks for the review @ParaskP7 and proceeding with these non-trivial test cases! 🙌
